core/lib/dal/src/tee_verifier_input_producer_dal.rs

```diff
@@ -0,0 +1,233 @@
+    InProgress,
+    /// The final (happy case) state we expect all jobs to end up. After the run is complete,
+    /// the job uploaded it's inputs, it lands in successful.
+    Successful,
+    /// The job failed for reasons. It will be marked as such and the error persisted in DB.
+    /// If it failed less than MAX_ATTEMPTs, the job will be retried,
+    /// otherwise it will stay in this state as final state.
+    Failed,
+}
+
+impl TeeVerifierInputProducerDal<'_, '_> {
+    pub async fn create_tee_verifier_input_producer_job(
+        &mut self,
+        l1_batch_number: L1BatchNumber,
+    ) -> DalResult<()> {
+        sqlx::query!(
+            r#"
+            INSERT INTO
+                tee_verifier_input_producer_jobs (l1_batch_number, status, created_at, updated_at)
+            VALUES
+                ($1, $2, NOW(), NOW())
+            ON CONFLICT (l1_batch_number) DO NOTHING
+            "#,
+            i64::from(l1_batch_number.0),
+            TeeVerifierInputProducerJobStatus::Queued as TeeVerifierInputProducerJobStatus,
+        )
+        .instrument("create_tee_verifier_input_producer_job")
+        .with_arg("l1_batch_number", &l1_batch_number)
+        .report_latency()
+        .execute(self.storage)
+        .await?;
+
+        Ok(())
+    }
+
+    pub async fn get_next_tee_verifier_input_producer_job(
+        &mut self,
+    ) -> DalResult<Option<L1BatchNumber>> {
+        let l1_batch_number = sqlx::query!(
+            r#"
+            UPDATE tee_verifier_input_producer_jobs
+            SET
+                status = $1,
+                attempts = attempts + 1,
+                updated_at = NOW(),
+                processing_started_at = NOW()
+            WHERE
+                l1_batch_number = (
+                    SELECT
+                        l1_batch_number
+                    FROM
+                        tee_verifier_input_producer_jobs
+                    WHERE
+                        status = $2
+                        OR (
+                            status = $1
+                            AND processing_started_at < NOW() - $4::INTERVAL
+                        )
+                        OR (
+                            status = $3
+                            AND attempts < $5
+                        )
+                    ORDER BY
+                        l1_batch_number ASC
+                    LIMIT
+                        1
+                    FOR UPDATE
+                        SKIP LOCKED
+                )
+            RETURNING
+                tee_verifier_input_producer_jobs.l1_batch_number
+            "#,
+            TeeVerifierInputProducerJobStatus::InProgress as TeeVerifierInputProducerJobStatus,
+            TeeVerifierInputProducerJobStatus::Queued as TeeVerifierInputProducerJobStatus,
+            TeeVerifierInputProducerJobStatus::Failed as TeeVerifierInputProducerJobStatus,
```

Comment:
why do we extract those to parameters? Makes it a little harder to read the query. Ah, or is it to prevent potential typos, using the type system for this? In theory sqlx should catch it when checking query against the DB

Reply:
it was copy pasta from https://github.com/matter-labs/zksync-era/blob/6de925c64ecc3c23bf16346ab56b171387b3fbba/core/lib/dal/src/basic_witness_input_producer_dal.rs#L114-L116